### PR TITLE
Template literal types for template literal expressions

### LINKED ISSUE #41631: Try to preserve template-ness of string literal types where possible
## Search Terms

template literal preserve

## Suggestion

Now that we have template literal types, I would love to have the ability to apply them to string template literals.

## Use Cases

This would be really useful in the [ioBroker type declarations](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/iobroker) where we try to infer the object types returned by some methods from the given IDs. These IDs can be directly specified in the method calls, but are in reality often built from multiple parts.

A dumbed down example is shown in the following examples:

## Examples

[Playground](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201122#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwygGsQBnAGSwxBiggB4AVeEAD2tWFPlIxi1QBzAHwAKCFRp0AXPEYBKWczYcu8AAaIcOAHQAjWDoAkAbwGIa8AEoBfdfAD81+LLRFUOAO6oA3AFgAKECwPF4CAEZ4AF4CYjJKaloIUQAiLV0DGH0oAC8U+R94AHoi+BSDPODQjHgsYAAmaLL07KyKlP8AkNQwjEaYwhIKSSTROvqC4tLy3JTAwNBIWARusNIcAFsQAGU+AUFZXn4hTtWajABmJsH4kbpRTW1W4xN1rd3jwTtJkpZWAAdwNRgIc9kIADTwKCYZAyeBuDzeeZdaq1YAAFiajwyhlMbx2YK+6lOqIwmIGcWGiXudXRP1KbEBmBAIJ4hMh0IwsIgrlQ7i8qECQA)

```ts
declare function takesLiteral<T extends string>(literal: T): T extends `foo.bar.${infer R}` ? R : unknown;

const t1 = takesLiteral("foo.bar.baz"); // "baz"
const id2 = "foo.bar.baz";
const t2 = takesLiteral(id2); // "baz"

declare const someString: string;
const t3 = takesLiteral(`foo.bar.${someString}`); // expected: string, actual: unknown

const id4 = `foo.bar.${someString}`;
const t4 = takesLiteral(id4); // expected: string, actual: unknown
```

Both `id4` and the string passed to the third call of `takesLiteral` are string constants that are created with a template literal. TypeScript should be able to preserve that *literal-ness*, since it already knows we're feeding a string into a template string.

This means we could work with literals of type `foo.bar.${string}`, not `string` like we currently do.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we infer template literal types for template literal expressions. The PR adheres to the same principles as those of literal types in #10676 and widening vs. non-widening literal types in #11126. For example:

```ts
const n: number = 123;
const s1 = `${n}px`;  // `${number}px`
const s2: `${number}px` = s1;
const s3: `${number}pt` = s1;  // Error
let v1 = s1;  // string (because of widening)
```

As it is generally a breaking change for us to more strongly type _anything_, I'm marking this as a breaking change. In practice, the breaks appear to be rare and easy to fix.

Fixes #41631.

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -4894,6 +4894,8 @@ namespace ts {
         Literal = StringLiteral | NumberLiteral | BigIntLiteral | BooleanLiteral,
         Unit = Literal | UniqueESSymbol | Nullable,
         StringOrNumberLiteral = StringLiteral | NumberLiteral,
+        StringLikeLiteral = StringLiteral | TemplateLiteral,
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts:4897]: I'd say mark this as `/** @internal */`

File: src/compiler/types.ts [hunk 2]
```diff
@@ -4894,6 +4894,8 @@ namespace ts {
         Literal = StringLiteral | NumberLiteral | BigIntLiteral | BooleanLiteral,
         Unit = Literal | UniqueESSymbol | Nullable,
         StringOrNumberLiteral = StringLiteral | NumberLiteral,
+        StringLikeLiteral = StringLiteral | TemplateLiteral,
+        FreshableLiteral = Literal | TemplateLiteral,
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts:4898]: Same here, mark internal

File: tests/baselines/reference/emitExponentiationOperatorInTempalteString4.types [hunk 3]
```diff
@@ -62,7 +62,7 @@ var s;
 >t1 : number
 
 `${typeof (t1 ** t2 ** t1) } world`;
->`${typeof (t1 ** t2 ** t1) } world` : string
+>`${typeof (t1 ** t2 ** t1) } world` : "string world" | "number world" | "bigint world" | "boolean world" | "symbol world" | "undefined world" | "object world" | "function world"
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/emitExponentiationOperatorInTempalteString4.types:65]: What an age to be alive.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
The implementation looks good apart from the API stuff.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by ahejlsberg** [comment]
Test suites revealed that template literal types didn't consistently have a `string` constraint and therefore weren't assignable to `String`, `Object`, `{}`, and so on. Latest commits fix that.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this

**Review by weswigham — APPROVED** [review]

**Comment by ahejlsberg** [comment]
DT run is clean. Performance looks fine. Latest RWC run shows a few errors that shouldn't have been errors disappearing.

**Comment by ahejlsberg** [comment]
Three user test suites have new errors.

### office-ui-fabric

We are now (correctly) reporting an error that was previously silenced because they're compiling with `strictFunctionTypes` off. Simplified example:

```ts
declare const someStringArray: string[];
someStringArray.map(key => !!true ? `foo${key}` : null).filter((s: string) => s);
```

We now report an error neither `` `foo${string` | null`` or `string` are assignable to each other. Previously it (unsoundly) succeeded because `string` is assignable to `string | null`.

We're also reporting an error on something similar to this:

```ts
declare const someStringArray: string[];
someStringArray.map(key => `foo${key}`).concat(someStringArray);  // string[] not assignable to `foo${string}`[]
```

Ideally this wouldn't be an error. The core issue is that `concat` types the array and the argument using the same type variable. It really should be `T` and `U` with the result typed as `(T | U)[]`.

### fp-ts

One new error that looks to be correct and simply a result of the stronger typing.

### webpack

One new error on something similar to this:

```ts
declare const someArray: unknown[];
someArray.map(item => `foo${item}`).unshift("someString");  // Type "someString" not assignable to type `foo${string}`
```

Not much we can do about this. The array is now of type `` `foo${string}`[]`` instead of just `string[]`, so it is an error to insert a value of type `"someString"`.

**Comment by ahejlsberg** [comment]
I'm going to mark this as a breaking change given what we're seeing in the user test suites.

**Comment by weswigham** [comment]
Oof, apparently breaky enough to break our own build post-LKG. Namely, we have
```ts
                    const internal = jsDocStyle ? `/**@internal*/` : `/*@internal*/`;
                    // ...
                    appendText(fs, sources[Project.second][Source.ts][Part.one], `
class normalC {
    ${internal} constructor() { }
    ${internal} prop: string;
    ${internal} method() { }
    ${internal} get c() { return 10; }
    ${internal} set c(val: number) { }
}
namespace normalN {
    ${internal} export class C { }
    ${internal} export function foo() {}
    ${internal} export namespace someNamespace { export class C {} }
    ${internal} export namespace someOther.something { export class someClass {} }
    ${internal} export import someImport = someNamespace.C;
    ${internal} export type internalType = internalC;
    ${internal} export const internalConst = 10;
    ${internal} export enum internalEnum { a, b, c }
}
${internal} class internalC {}
${internal} function internalfoo() {}
${internal} namespace internalNamespace { export class someClass {} }
${internal} namespace internalOther.something { export class someClass {} }
${internal} import internalImport = internalNamespace.someClass;
${internal} type internalType = internalC;
${internal} const internalConst = 10;
${internal} enum internalEnum { a, b, c }`);
```
in our test harness.

**Comment by weswigham** [comment]
https://github.com/microsoft/TypeScript/pull/41942 is up with a fix to the build break, assuming we don't want to do anything with this change,

**Comment by jcalz** [comment]
So #42588 undoes #41891, right?  Just cross-linking them.

**Comment by pedrodurek** [comment]
Hey @DanielRosenwasser, considering that changes here were rolled back, do we have a new issue opened for this? Or you'll reopen this one https://github.com/microsoft/TypeScript/issues/41631?

**Pull request merged**